FuelLabs/fuel-indexer
# Add postgres constraints

### Description
- This PR does a few things:
  - Implements `indexed` directives
    - The directives were pre-existing, this PR just links them to the Postgres layer
  - Adds foreign key support
  - Updates the Dockerfile rust version to `1.62`
 
- Keep in mind the world of Postgres constraints is vast and complex, this PR attempts to implement the basic/rudimentary constraint usage that covers 85% of use cases. We can always add more later.  
- Checkout the below linked issues to see some of the constraints mentioned in order to get this work out the door
- Addresses:
  - https://github.com/FuelLabs/fuel-indexer/issues/133
  - https://github.com/FuelLabs/fuel-indexer/issues/112
 
### Testing steps
- [ ] Create a schema file and use the `@indexed` directive on some of the fields
  - After starting the services, to into Postgres and see your table's schema. You should see the indices
- [ ] Create a foreign key relationship
  - You should see the constraints added after starting the service and inspecting the Postgres schema

```graphql
# Example foreign key schema
schema {
    query: QueryRoot
}

type QueryRoot {
    count: Count
    another_count: AnotherCount
}

type Count {
    id: ID!
    count: UInt8!
    timestamp: UInt8!
}


type AnotherCount {
    id: ID!
    count: UInt8!
    timestamp: UInt8!
    address: Address! @indexed
    reference_count: Count!
}


type YetAnotherCount {
    id: ID!
    count: UInt8!
    timestamp: UInt8!
    address: Address! @indexed
    another_reference_count: Count!
}

```

### Notes
- This **does** handle one-to-many relationships when querying
- Querying these relationships and serializing entities is being tracked in https://github.com/FuelLabs/fuel-indexer/issues/136
- Read the issues to understand what constraints were put in place for the sake of simplicity
- Indices limited to `btree`
- Indices limited to a single field with no clauses
- Foreign keys limited to a single field with one-to-many and one-to-one support (no many-to-many)
- Foreign keys only support `NO ACTION` on delete and update
